# *Data Preparation with Python* - Bagian 1
Pertanyaan, koreksi, maupun saran dapat disampaikan ke galuh.tunggadewi[at]gmail.com atau disampaikan dengan cara membuka [isu baru di GitHub](https://github.com/galuhsahid/data-preparation-with-python). Lisensi dapat dilihat [di sini](https://creativecommons.org/licenses/by-nc-sa/4.0/).

In [18]:
import pandas as pd

# Membaca data

In [19]:
df = pd.read_csv("data/penduduk_telepon_seluler_pedesaan.csv")

📖  Dokumentasi `pandas.read_csv`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)

- Untuk membaca file `json`:  [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_json.html)

In [20]:
df

nama_series;kode_wilayah;wilayah;tahun;nilai;target
0    Proporsi Penduduk yang Memiliki Telepon Selule... 
1    Proporsi Penduduk yang Memiliki Telepon Selule... 
2    Proporsi Penduduk yang Memiliki Telepon Selule... 
3    Proporsi Penduduk yang Memiliki Telepon Selule... 
4    Proporsi Penduduk yang Memiliki Telepon Selule... 
5    Proporsi Penduduk yang Memiliki Telepon Selule... 
6    Proporsi Penduduk yang Memiliki Telepon Selule... 
7    Proporsi Penduduk yang Memiliki Telepon Selule... 
8    Proporsi Penduduk yang Memiliki Telepon Selule... 
9    Proporsi Penduduk yang Memiliki Telepon Selule... 
10   Proporsi Penduduk yang Memiliki Telepon Selule... 
11   Proporsi Penduduk yang Memiliki Telepon Selule... 
12   Proporsi Penduduk yang Memiliki Telepon Selule... 
13   Proporsi Penduduk yang Memiliki Telepon Selule... 
14   Proporsi Penduduk yang Memiliki Telepon Selule... 
15   Proporsi Penduduk yang Memiliki Telepon Selule... 
16   Proporsi Penduduk yang Memiliki Telepon Selule... 
17   Proporsi Penduduk yang Memiliki Telepon Selule... 
18   Proporsi Penduduk yang Memiliki Telepon Selule... 
19   Proporsi Penduduk yang Memiliki Telepon Selule... 
20   Proporsi Penduduk yang Memiliki Telepon Selule... 
21   Proporsi Penduduk yang Memiliki Telepon Selule... 
22   Proporsi Penduduk yang Memiliki Telepon Selule... 
23   Proporsi Penduduk yang Memiliki Telepon Selule... 
24   Proporsi Penduduk yang Memiliki Telepon Selule... 
25   Proporsi Penduduk yang Memiliki Telepon Selule... 
26   Proporsi Penduduk yang Memiliki Telepon Selule... 
27   Proporsi Penduduk yang Memiliki Telepon Selule... 
28   Proporsi Penduduk yang Memiliki Telepon Selule... 
29   Proporsi Penduduk yang Memiliki Telepon Selule... 
..                                                 ... 
495  Proporsi Penduduk yang Memiliki Telepon Selule... 
496  Proporsi Penduduk yang Memiliki Telepon Selule... 
497  Proporsi Penduduk yang Memiliki Telepon Selule... 
498  Proporsi Penduduk yang Memiliki Telepon Selule... 
499  Proporsi Penduduk yang Memiliki Telepon Selule... 
500  Proporsi Penduduk yang Memiliki Telepon Selule... 
501  Proporsi Penduduk yang Memiliki Telepon Selule... 
502  Proporsi Penduduk yang Memiliki Telepon Selule... 
503  Proporsi Penduduk yang Memiliki Telepon Selule... 
504  Proporsi Penduduk yang Memiliki Telepon Selule... 
505  Proporsi Penduduk yang Memiliki Telepon Selule... 
506  Proporsi Penduduk yang Memiliki Telepon Selule... 
507  Proporsi Penduduk yang Memiliki Telepon Selule... 
508  Proporsi Penduduk yang Memiliki Telepon Selule... 
509  Proporsi Penduduk yang Memiliki Telepon Selule... 
510  Proporsi Penduduk yang Memiliki Telepon Selule... 
511  Proporsi Penduduk yang Memiliki Telepon Selule... 
512  Proporsi Penduduk yang Memiliki Telepon Selule... 
513  Proporsi Penduduk yang Memiliki Telepon Selule... 
514  Proporsi Penduduk yang Memiliki Telepon Selule... 
515  Proporsi Penduduk yang Memiliki Telepon Selule... 
516  Proporsi Penduduk yang Memiliki Telepon Selule... 
517  Proporsi Penduduk yang Memiliki Telepon Selule... 
518  Proporsi Penduduk yang Memiliki Telepon Selule... 
519  Proporsi Penduduk yang Memiliki Telepon Selule... 
520  Proporsi Penduduk yang Memiliki Telepon Selule... 
521  Proporsi Penduduk yang Memiliki Telepon Selule... 
522  Proporsi Penduduk yang Memiliki Telepon Selule... 
523  Proporsi Penduduk yang Memiliki Telepon Selule... 
524  Proporsi Penduduk yang Memiliki Telepon Selule... 

[525 rows x 1 columns]

Dalam sebuah *file* CSV, masing-masing *field* tidak harus dipisahkan oleh tanda `,` (koma), melainkan bisa dengan pemisah-pemisah seperti:
- `;` (titik koma)
- `\t` (*tab*)
- `|` (*pipe*)

Pada contoh di atas, pandas tidak membaca data kita dengan benar. Hal ini karena tanda baca yang memisahkan kolom-kolom di *file* yang dibaca adalah tanda titik koma. Apabila kita tidak mendefinisikan secara eksplisit separator apa yang harus digunakan, maka secara *default* pandas akan menganggap bahwa separator yang digunakan adalah tanda koma.

In [86]:
df = pd.read_csv("data/penduduk_telepon_seluler_pedesaan.csv", sep=";")

Apabila kita ingin melihat data yang telah kita *load*, namun kita tidak perlu melihat semua datanya, kita dapat menggunakan `head()` yang akan memunculkan lima baris pertama:

In [87]:
df.head()

nama_series  kode_wilayah    wilayah  \
0  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
1  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
2  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
3  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
4  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   

   tahun  nilai  target  
0   2001    NaN   100.0  
1   2002    NaN   100.0  
2   2003    NaN   100.0  
3   2004    NaN   100.0  
4   2005   8.19   100.0

Apabila kita butuh untuk melihat lebih dari lima baris, atau bahkan kurang dari lima baris, kita dapat memasukkan angka jumlah baris yang ingin kita lihat:

In [88]:
df.head(10)

nama_series  kode_wilayah    wilayah  \
0  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
1  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
2  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
3  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
4  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
5  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
6  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
7  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
8  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
9  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   

   tahun  nilai  target  
0   2001    NaN   100.0  
1   2002    NaN   100.0  
2   2003    NaN   100.0  
3   2004    NaN   100.0  
4   2005   8.19   100.0  
5   2006  12.27   100.0  
6   2007  24.33   100.0  
7   2008  38.15   100.0  
8   2009  49.21   100.0  
9   2010  61.01   100.0

Kita juga dapat menggunakan `tail()` apabila ingin melihat baris-baris yang ada di urutan bawah.

In [89]:
df.tail()

nama_series  kode_wilayah wilayah  \
520  Proporsi Penduduk yang Memiliki Telepon Selule...            94   Papua   
521  Proporsi Penduduk yang Memiliki Telepon Selule...            94   Papua   
522  Proporsi Penduduk yang Memiliki Telepon Selule...            94   Papua   
523  Proporsi Penduduk yang Memiliki Telepon Selule...            94   Papua   
524  Proporsi Penduduk yang Memiliki Telepon Selule...            94   Papua   

     tahun  nilai  target  
520   2011  17.26     NaN  
521   2012  20.50     NaN  
522   2013  25.08     NaN  
523   2014  26.31     NaN  
524   2015  32.51     NaN

Struktur data di pandas secara garis besar diimplementasikan sebagai salah satu dari dua kelas berikut:
- `DataFrame`: bisa dianggap sebagai sebuah tabel data relasional dengan baris-baris (*rows*) dan kolom-kolom (*columns*)
- `Series`: sebuah kolom. Sebuah DataFrame memiliki satu atau lebih `Series`.

![Series vs DataFrame](https://storage.googleapis.com/lds-media/images/series-and-dataframe.width-1200.png)
[*Source*](https://www.learndatasci.com/tutorials/python-pandas-tutorial-complete-introduction-for-beginners/)

Dengan menggunakan `type()`, fungsi dari Python yang dapat digunakan untuk mengecek tipe dari sebuah variabel, kita dapat melihat bahwa tipe dari variabel `df` adalah DataFrame.

In [90]:
type(df)

pandas.core.frame.DataFrame

# Mengakses data

## Mengakses kolom

In [91]:
df["wilayah"]

0        Indonesia
1        Indonesia
2        Indonesia
3        Indonesia
4        Indonesia
5        Indonesia
6        Indonesia
7        Indonesia
8        Indonesia
9        Indonesia
10       Indonesia
11       Indonesia
12       Indonesia
13       Indonesia
14       Indonesia
15            Aceh
16            Aceh
17            Aceh
18            Aceh
19            Aceh
20            Aceh
21            Aceh
22            Aceh
23            Aceh
24            Aceh
25            Aceh
26            Aceh
27            Aceh
28            Aceh
29            Aceh
          ...     
495    Papua Barat
496    Papua Barat
497    Papua Barat
498    Papua Barat
499    Papua Barat
500    Papua Barat
501    Papua Barat
502    Papua Barat
503    Papua Barat
504    Papua Barat
505    Papua Barat
506    Papua Barat
507    Papua Barat
508    Papua Barat
509    Papua Barat
510          Papua
511          Papua
512          Papua
513          Papua
514          Papua
515          Papua
516         

Dengan menggunakan `type()`, kita dapat melihat bahwa tipe dari kolom `df["wilayah"]` adalah Series.

In [27]:
type(df["wilayah"])

pandas.core.series.Series

Melihat nilai-nilai unik dari kolom **wilayah**:

In [92]:
df["wilayah"].unique()

array(['Indonesia', 'Aceh', 'Sumatera Utara', 'Sumatera Barat', 'Riau',
       'Jambi', 'Sumatera Selatan', 'Bengkulu', 'Lampung',
       'Kepulauan Bangka Belitung', 'Kepulauan Riau', 'DKI Jakarta',
       'Jawa Barat', 'Jawa Tengah', 'DI Yogyakarta', 'Jawa Timur',
       'Banten', 'Bali', 'Nusa Tenggara Barat', 'Nusa Tenggara Timur',
       'Kalimantan Barat', 'Kalimantan Tengah', 'Kalimantan Selatan',
       'Kalimantan Timur', 'Kalimantan Utara', 'Sulawesi Utara',
       'Sulawesi Tengah', 'Sulawesi Selatan', 'Sulawesi Tenggara',
       'Gorontalo', 'Sulawesi Barat', 'Maluku', 'Maluku Utara',
       'Papua Barat', 'Papua'], dtype=object)

Apabila kita ingin mengecek kolom apa saja yang ada di DataFrame kita:

In [93]:
df.columns

Index(['nama_series', 'kode_wilayah', 'wilayah', 'tahun', 'nilai', 'target'], dtype='object')

## Mengakses baris

- `iloc` - *input* berupa posisi (hanya menerima integer)
- `loc` -  *input* berupa label (tidak harus integer, bisa saja string apabila label-nya merupakan sebuah string)
- `ix` - *deprecated*

### iloc

Untuk mendemonstrasikan perbedaan antara `loc` dan `iloc`, kita akan membuat satu DataFrame baru di variabel `df_shuffled` yang merupakan DataFrame `df` yang urutannya diacak:

In [96]:
import numpy as np

df_shuffled = df.reindex(np.random.permutation(df.index))

In [97]:
df.head(3)

nama_series  kode_wilayah    wilayah  \
0  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
1  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
2  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   

   tahun  nilai  target  
0   2001    NaN   100.0  
1   2002    NaN   100.0  
2   2003    NaN   100.0

In [98]:
df_shuffled.head(3)

nama_series  kode_wilayah  \
261  Proporsi Penduduk yang Memiliki Telepon Selule...            51   
310  Proporsi Penduduk yang Memiliki Telepon Selule...            61   
414  Proporsi Penduduk yang Memiliki Telepon Selule...            73   

              wilayah  tahun  nilai  target  
261              Bali   2007  38.60     NaN  
310  Kalimantan Barat   2011  72.65     NaN  
414  Sulawesi Selatan   2010  70.20     NaN

Dengan kode di bawah ini, kita akan mengambil data baris pertama dari DataFrame `df_shuffled`:

In [99]:
df_shuffled.iloc[0]

nama_series     Proporsi Penduduk yang Memiliki Telepon Selule...
kode_wilayah                                                   51
wilayah                                                      Bali
tahun                                                        2007
nilai                                                        38.6
target                                                        NaN
Name: 261, dtype: object

⚠️ Python menghitung dari 0, sehingga apabila kita ingin mengambil data di baris pertama, maka posisinya adalah posisi ke-0.

Meskipun indeks-nya adalah 261, namun data tersebut tetap terambil karena **posisi**-nya ada di baris pertama DataFrame. Apabila kita mencoba untuk mengambil data di baris ke-261 (posisi ke-260), yang muncul adalah baris di posisi tersebut, bukan baris dengan indeks 261.

In [100]:
df_shuffled.iloc[260]

nama_series     Proporsi Penduduk yang Memiliki Telepon Selule...
kode_wilayah                                                   94
wilayah                                                     Papua
tahun                                                        2005
nilai                                                        4.61
target                                                        NaN
Name: 514, dtype: object

Pada contoh di atas, baris yang muncul adalah baris dengan indeks 514. Ternyata, setelah diacak, baris dengan indeks 514 menempati baris ke-261.

### loc

`loc` mengambil baris berdasarkan indeks.

In [104]:
df_shuffled.head(3)

nama_series  kode_wilayah  \
261  Proporsi Penduduk yang Memiliki Telepon Selule...            51   
310  Proporsi Penduduk yang Memiliki Telepon Selule...            61   
414  Proporsi Penduduk yang Memiliki Telepon Selule...            73   

              wilayah  tahun  nilai  target  
261              Bali   2007  38.60     NaN  
310  Kalimantan Barat   2011  72.65     NaN  
414  Sulawesi Selatan   2010  70.20     NaN

In [105]:
df_shuffled.loc[0]

nama_series     Proporsi Penduduk yang Memiliki Telepon Selule...
kode_wilayah                                                    1
wilayah                                                 Indonesia
tahun                                                        2001
nilai                                                         NaN
target                                                        100
Name: 0, dtype: object

Pada contoh di atas, kita tidak mendapatkan data di baris ke-0, melainkan baris dengan indeks 0.

Apabila kita ingin mendapatkan baris dengan indeks 261:

In [108]:
df_shuffled.loc[261]

nama_series     Proporsi Penduduk yang Memiliki Telepon Selule...
kode_wilayah                                                   51
wilayah                                                      Bali
tahun                                                        2007
nilai                                                        38.6
target                                                        NaN
Name: 261, dtype: object

⚠️ Indeks bukan merupakan posisi. Baris dengan indeks 261 akan benar-benar memiliki indeks 261, sehingga tidak perlu dikurangi lagi. Apabila kita coba kode berikut:

In [109]:
df_shuffled.loc[260]

nama_series     Proporsi Penduduk yang Memiliki Telepon Selule...
kode_wilayah                                                   51
wilayah                                                      Bali
tahun                                                        2006
nilai                                                       27.22
target                                                        NaN
Name: 260, dtype: object

Kita akan mendapatkan baris dengan indeks 260.

### Latihan
Bagaimana cara agar kita bisa mendapatkan data di kolom "nilai" yang ada di baris kelima dari dataframe `df` (yang belum di-*shuffle*)?

In [110]:
df.head()

nama_series  kode_wilayah    wilayah  \
0  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
1  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
2  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
3  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
4  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   

   tahun  nilai  target  
0   2001    NaN   100.0  
1   2002    NaN   100.0  
2   2003    NaN   100.0  
3   2004    NaN   100.0  
4   2005   8.19   100.0

#### Solusi

Ingat bahwa di Python, indeks dimulai dari hitungan 0, jadi untuk mengambil baris kelima kita harus menggunakan indeks = 4.

In [111]:
df.iloc[4]

nama_series     Proporsi Penduduk yang Memiliki Telepon Selule...
kode_wilayah                                                    1
wilayah                                                 Indonesia
tahun                                                        2005
nilai                                                        8.19
target                                                        100
Name: 4, dtype: object

In [112]:
df["nilai"].iloc[4]

8.19

## Mengakses sebagian data (range)

In [113]:
["wilayah", "tahun", "nilai"]

['wilayah', 'tahun', 'nilai']

In [114]:
subset_df = df[["wilayah", "tahun", "nilai"]]

In [115]:
subset_df.head()

wilayah  tahun  nilai
0  Indonesia   2001    NaN
1  Indonesia   2002    NaN
2  Indonesia   2003    NaN
3  Indonesia   2004    NaN
4  Indonesia   2005   8.19

In [116]:
selected_columns = ["wilayah", "tahun"]

In [117]:
subset_df_2 = df[selected_columns]

In [118]:
subset_df_2.head()

wilayah  tahun
0  Indonesia   2001
1  Indonesia   2002
2  Indonesia   2003
3  Indonesia   2004
4  Indonesia   2005

Mengambil baris-baris dengan indeks 2, 0, dan 10:

In [119]:
subset_rows = df.loc[[2, 0, 10]]

In [120]:
subset_rows.head()

nama_series  kode_wilayah  \
2   Proporsi Penduduk yang Memiliki Telepon Selule...             1   
0   Proporsi Penduduk yang Memiliki Telepon Selule...             1   
10  Proporsi Penduduk yang Memiliki Telepon Selule...             1   

      wilayah  tahun  nilai  target  
2   Indonesia   2003    NaN   100.0  
0   Indonesia   2001    NaN   100.0  
10  Indonesia   2011  69.46   100.0

In [50]:
subset = df.loc[:, selected_columns]

In [51]:
subset

wilayah  tahun
0      Indonesia   2001
1      Indonesia   2002
2      Indonesia   2003
3      Indonesia   2004
4      Indonesia   2005
5      Indonesia   2006
6      Indonesia   2007
7      Indonesia   2008
8      Indonesia   2009
9      Indonesia   2010
10     Indonesia   2011
11     Indonesia   2012
12     Indonesia   2013
13     Indonesia   2014
14     Indonesia   2015
15          Aceh   2001
16          Aceh   2002
17          Aceh   2003
18          Aceh   2004
19          Aceh   2005
20          Aceh   2006
21          Aceh   2007
22          Aceh   2008
23          Aceh   2009
24          Aceh   2010
25          Aceh   2011
26          Aceh   2012
27          Aceh   2013
28          Aceh   2014
29          Aceh   2015
..           ...    ...
495  Papua Barat   2001
496  Papua Barat   2002
497  Papua Barat   2003
498  Papua Barat   2004
499  Papua Barat   2005
500  Papua Barat   2006
501  Papua Barat   2007
502  Papua Barat   2008
503  Papua Barat   2009
504  Papua Barat   2010
505  Papua Barat   2011
506  Papua Barat   2012
507  Papua Barat   2013
508  Papua Barat   2014
509  Papua Barat   2015
510        Papua   2001
511        Papua   2002
512        Papua   2003
513        Papua   2004
514        Papua   2005
515        Papua   2006
516        Papua   2007
517        Papua   2008
518        Papua   2009
519        Papua   2010
520        Papua   2011
521        Papua   2012
522        Papua   2013
523        Papua   2014
524        Papua   2015

[525 rows x 2 columns]

In [53]:
subset = df.loc[3:5, selected_columns]

In [54]:
subset

wilayah  tahun
3  Indonesia   2004
4  Indonesia   2005
5  Indonesia   2006

Memunculkan seluruh baris dengan tahun 2004:

In [58]:
subset = df.loc[df["tahun"] == 2004, selected_columns]

In [59]:
subset.head()

wilayah  tahun
3        Indonesia   2004
18            Aceh   2004
33  Sumatera Utara   2004
48  Sumatera Barat   2004
63            Riau   2004

Pastikan bahwa di dalam *dataframe* kita, hanya ada tahun 2004:

In [60]:
subset["tahun"].unique()

array([2004])

In [61]:
subset["tahun"].nunique()

1

Memunculkan baris-baris dengan *field* "nilai" yang kosong:

In [62]:
df[df["nilai"].isnull()].head()

nama_series  kode_wilayah  \
0   Proporsi Penduduk yang Memiliki Telepon Selule...             1   
1   Proporsi Penduduk yang Memiliki Telepon Selule...             1   
2   Proporsi Penduduk yang Memiliki Telepon Selule...             1   
3   Proporsi Penduduk yang Memiliki Telepon Selule...             1   
15  Proporsi Penduduk yang Memiliki Telepon Selule...            11   

      wilayah  tahun  nilai  target  
0   Indonesia   2001    NaN   100.0  
1   Indonesia   2002    NaN   100.0  
2   Indonesia   2003    NaN   100.0  
3   Indonesia   2004    NaN   100.0  
15       Aceh   2001    NaN     NaN

Memunculkan baris-baris dengan *field* "nilai" yang tidak kosong:

In [63]:
df[~df["nilai"].isnull()].head()

nama_series  kode_wilayah    wilayah  \
4  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
5  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
6  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
7  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   
8  Proporsi Penduduk yang Memiliki Telepon Selule...             1  Indonesia   

   tahun  nilai  target  
4   2005   8.19   100.0  
5   2006  12.27   100.0  
6   2007  24.33   100.0  
7   2008  38.15   100.0  
8   2009  49.21   100.0

### Latihan
1. Bagaimana caranya agar kita dapat mengambil data dari baris ke-400 sampai baris terakhir untuk kolom wilayah, tahun, target?
2. Bagaimana kita bisa mendapatkan baris untuk tahun 2004, 2005, dan 2007 untuk kolom wilayah tahun, target?
3. Bagaimana kita bisa mendapatkan baris untuk wilayah-wilayah yang ada di pulau Kalimantan di tahun 2005 - 2008?


#### Hint

Untuk nomor 2, kita bisa menggunakan operator logika:
- `&` untuk `dan`
- `|` untuk `atau`

#### Solusi 1

In [65]:
selected_cols = ["wilayah", "tahun", "target"]
subset = df[selected_cols].iloc[399:]

In [66]:
subset

wilayah  tahun  target
399    Sulawesi Tengah   2010     NaN
400    Sulawesi Tengah   2011     NaN
401    Sulawesi Tengah   2012     NaN
402    Sulawesi Tengah   2013     NaN
403    Sulawesi Tengah   2014     NaN
404    Sulawesi Tengah   2015     NaN
405   Sulawesi Selatan   2001     NaN
406   Sulawesi Selatan   2002     NaN
407   Sulawesi Selatan   2003     NaN
408   Sulawesi Selatan   2004     NaN
409   Sulawesi Selatan   2005     NaN
410   Sulawesi Selatan   2006     NaN
411   Sulawesi Selatan   2007     NaN
412   Sulawesi Selatan   2008     NaN
413   Sulawesi Selatan   2009     NaN
414   Sulawesi Selatan   2010     NaN
415   Sulawesi Selatan   2011     NaN
416   Sulawesi Selatan   2012     NaN
417   Sulawesi Selatan   2013     NaN
418   Sulawesi Selatan   2014     NaN
419   Sulawesi Selatan   2015     NaN
420  Sulawesi Tenggara   2001     NaN
421  Sulawesi Tenggara   2002     NaN
422  Sulawesi Tenggara   2003     NaN
423  Sulawesi Tenggara   2004     NaN
424  Sulawesi Tenggara   2005     NaN
425  Sulawesi Tenggara   2006     NaN
426  Sulawesi Tenggara   2007     NaN
427  Sulawesi Tenggara   2008     NaN
428  Sulawesi Tenggara   2009     NaN
..                 ...    ...     ...
495        Papua Barat   2001     NaN
496        Papua Barat   2002     NaN
497        Papua Barat   2003     NaN
498        Papua Barat   2004     NaN
499        Papua Barat   2005     NaN
500        Papua Barat   2006     NaN
501        Papua Barat   2007     NaN
502        Papua Barat   2008     NaN
503        Papua Barat   2009     NaN
504        Papua Barat   2010     NaN
505        Papua Barat   2011     NaN
506        Papua Barat   2012     NaN
507        Papua Barat   2013     NaN
508        Papua Barat   2014     NaN
509        Papua Barat   2015     NaN
510              Papua   2001     NaN
511              Papua   2002     NaN
512              Papua   2003     NaN
513              Papua   2004     NaN
514              Papua   2005     NaN
515              Papua   2006     NaN
516              Papua   2007     NaN
517              Papua   2008     NaN
518              Papua   2009     NaN
519              Papua   2010     NaN
520              Papua   2011     NaN
521              Papua   2012     NaN
522              Papua   2013     NaN
523              Papua   2014     NaN
524              Papua   2015     NaN

[126 rows x 3 columns]

#### Solusi 2

In [67]:
subset = df.loc[((df["tahun"] == 2004) | (df["tahun"] == 2005) | (df["tahun"] == 2007)), selected_cols]

In [68]:
subset

wilayah  tahun  target
3                    Indonesia   2004   100.0
4                    Indonesia   2005   100.0
6                    Indonesia   2007   100.0
18                        Aceh   2004     NaN
19                        Aceh   2005     NaN
21                        Aceh   2007     NaN
33              Sumatera Utara   2004     NaN
34              Sumatera Utara   2005     NaN
36              Sumatera Utara   2007     NaN
48              Sumatera Barat   2004     NaN
49              Sumatera Barat   2005     NaN
51              Sumatera Barat   2007     NaN
63                        Riau   2004     NaN
64                        Riau   2005     NaN
66                        Riau   2007     NaN
78                       Jambi   2004     NaN
79                       Jambi   2005     NaN
81                       Jambi   2007     NaN
93            Sumatera Selatan   2004     NaN
94            Sumatera Selatan   2005     NaN
96            Sumatera Selatan   2007     NaN
108                   Bengkulu   2004     NaN
109                   Bengkulu   2005     NaN
111                   Bengkulu   2007     NaN
123                    Lampung   2004     NaN
124                    Lampung   2005     NaN
126                    Lampung   2007     NaN
138  Kepulauan Bangka Belitung   2004     NaN
139  Kepulauan Bangka Belitung   2005     NaN
141  Kepulauan Bangka Belitung   2007     NaN
..                         ...    ...     ...
378             Sulawesi Utara   2004     NaN
379             Sulawesi Utara   2005     NaN
381             Sulawesi Utara   2007     NaN
393            Sulawesi Tengah   2004     NaN
394            Sulawesi Tengah   2005     NaN
396            Sulawesi Tengah   2007     NaN
408           Sulawesi Selatan   2004     NaN
409           Sulawesi Selatan   2005     NaN
411           Sulawesi Selatan   2007     NaN
423          Sulawesi Tenggara   2004     NaN
424          Sulawesi Tenggara   2005     NaN
426          Sulawesi Tenggara   2007     NaN
438                  Gorontalo   2004     NaN
439                  Gorontalo   2005     NaN
441                  Gorontalo   2007     NaN
453             Sulawesi Barat   2004     NaN
454             Sulawesi Barat   2005     NaN
456             Sulawesi Barat   2007     NaN
468                     Maluku   2004     NaN
469                     Maluku   2005     NaN
471                     Maluku   2007     NaN
483               Maluku Utara   2004     NaN
484               Maluku Utara   2005     NaN
486               Maluku Utara   2007     NaN
498                Papua Barat   2004     NaN
499                Papua Barat   2005     NaN
501                Papua Barat   2007     NaN
513                      Papua   2004     NaN
514                      Papua   2005     NaN
516                      Papua   2007     NaN

[105 rows x 3 columns]

In [69]:
subset = df.loc[df["tahun"].isin([2004, 2005, 2007]), selected_cols]

In [70]:
subset.head()

wilayah  tahun  target
3   Indonesia   2004   100.0
4   Indonesia   2005   100.0
6   Indonesia   2007   100.0
18       Aceh   2004     NaN
19       Aceh   2005     NaN

#### Solusi 3

In [71]:
wilayah = ["Kalimantan Barat", "Kalimantan Tengah", "Kalimantan Timur", "Kalimantan Utara"]
tahun = [2004, 2005, 2006, 2007]

df[df["wilayah"].isin(wilayah) & df["tahun"].isin(tahun)]

nama_series  kode_wilayah  \
303  Proporsi Penduduk yang Memiliki Telepon Selule...            61   
304  Proporsi Penduduk yang Memiliki Telepon Selule...            61   
305  Proporsi Penduduk yang Memiliki Telepon Selule...            61   
306  Proporsi Penduduk yang Memiliki Telepon Selule...            61   
318  Proporsi Penduduk yang Memiliki Telepon Selule...            62   
319  Proporsi Penduduk yang Memiliki Telepon Selule...            62   
320  Proporsi Penduduk yang Memiliki Telepon Selule...            62   
321  Proporsi Penduduk yang Memiliki Telepon Selule...            62   
348  Proporsi Penduduk yang Memiliki Telepon Selule...            64   
349  Proporsi Penduduk yang Memiliki Telepon Selule...            64   
350  Proporsi Penduduk yang Memiliki Telepon Selule...            64   
351  Proporsi Penduduk yang Memiliki Telepon Selule...            64   
363  Proporsi Penduduk yang Memiliki Telepon Selule...            65   
364  Proporsi Penduduk yang Memiliki Telepon Selule...            65   
365  Proporsi Penduduk yang Memiliki Telepon Selule...            65   
366  Proporsi Penduduk yang Memiliki Telepon Selule...            65   

               wilayah  tahun  nilai  target  
303   Kalimantan Barat   2004    NaN     NaN  
304   Kalimantan Barat   2005   6.29     NaN  
305   Kalimantan Barat   2006  10.92     NaN  
306   Kalimantan Barat   2007  26.73     NaN  
318  Kalimantan Tengah   2004    NaN     NaN  
319  Kalimantan Tengah   2005   4.21     NaN  
320  Kalimantan Tengah   2006   8.83     NaN  
321  Kalimantan Tengah   2007  24.68     NaN  
348   Kalimantan Timur   2004    NaN     NaN  
349   Kalimantan Timur   2005  14.82     NaN  
350   Kalimantan Timur   2006  28.78     NaN  
351   Kalimantan Timur   2007  43.15     NaN  
363   Kalimantan Utara   2004    NaN     NaN  
364   Kalimantan Utara   2005    NaN     NaN  
365   Kalimantan Utara   2006    NaN     NaN  
366   Kalimantan Utara   2007    NaN     NaN

# Mengeksplor Data
Langkah ini penting karena ketika kita memasukkan data baru, sering kali pandas membacanya dengan cara yang tidak sesuai dengan yang kita inginkan (misalnya ada baris yang terpotong, atau separator-nya tidak tepat).

In [72]:
df.shape

(525, 6)

In [73]:
df.describe()

kode_wilayah        tahun       nilai  target
count    525.000000   525.000000  361.000000    15.0
mean      46.228571  2008.000000   53.558172   100.0
std       26.902161     4.324614   28.889846     0.0
min        1.000000  2001.000000    1.270000   100.0
25%       18.000000  2004.000000   25.870000   100.0
50%       51.000000  2008.000000   60.890000   100.0
75%       72.000000  2012.000000   80.230000   100.0
max       94.000000  2015.000000   94.050000   100.0

In [77]:
df["wilayah"].describe()

count            525
unique            35
top       Jawa Barat
freq              15
Name: wilayah, dtype: object

In [78]:
df["wilayah"].unique()

array(['Indonesia', 'Aceh', 'Sumatera Utara', 'Sumatera Barat', 'Riau',
       'Jambi', 'Sumatera Selatan', 'Bengkulu', 'Lampung',
       'Kepulauan Bangka Belitung', 'Kepulauan Riau', 'DKI Jakarta',
       'Jawa Barat', 'Jawa Tengah', 'DI Yogyakarta', 'Jawa Timur',
       'Banten', 'Bali', 'Nusa Tenggara Barat', 'Nusa Tenggara Timur',
       'Kalimantan Barat', 'Kalimantan Tengah', 'Kalimantan Selatan',
       'Kalimantan Timur', 'Kalimantan Utara', 'Sulawesi Utara',
       'Sulawesi Tengah', 'Sulawesi Selatan', 'Sulawesi Tenggara',
       'Gorontalo', 'Sulawesi Barat', 'Maluku', 'Maluku Utara',
       'Papua Barat', 'Papua'], dtype=object)

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525 entries, 0 to 524
Data columns (total 6 columns):
nama_series     525 non-null object
kode_wilayah    525 non-null int64
wilayah         525 non-null object
tahun           525 non-null int64
nilai           361 non-null float64
target          15 non-null float64
dtypes: float64(2), int64(2), object(2)
memory usage: 24.7+ KB


Tipe-tipe data di pandas:

![Pandas data type](https://pbpython.com/images/pandas_dtypes.png)
[*Source*](https://pbpython.com/pandas_dtypes.html)

## Mengganti tipe kolom

In [80]:
df["tahun"] = df["tahun"].astype("str")

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525 entries, 0 to 524
Data columns (total 6 columns):
nama_series     525 non-null object
kode_wilayah    525 non-null int64
wilayah         525 non-null object
tahun           525 non-null object
nilai           361 non-null float64
target          15 non-null float64
dtypes: float64(2), int64(1), object(3)
memory usage: 24.7+ KB


In [83]:
df = pd.read_csv("data/penduduk_telepon_seluler_pedesaan.csv", 
                 sep=";",
                 dtype={"tahun": str, "kode_wilayah": str})

In [84]:
df.head()

nama_series kode_wilayah    wilayah  \
0  Proporsi Penduduk yang Memiliki Telepon Selule...            1  Indonesia   
1  Proporsi Penduduk yang Memiliki Telepon Selule...            1  Indonesia   
2  Proporsi Penduduk yang Memiliki Telepon Selule...            1  Indonesia   
3  Proporsi Penduduk yang Memiliki Telepon Selule...            1  Indonesia   
4  Proporsi Penduduk yang Memiliki Telepon Selule...            1  Indonesia   

  tahun  nilai  target  
0  2001    NaN   100.0  
1  2002    NaN   100.0  
2  2003    NaN   100.0  
3  2004    NaN   100.0  
4  2005   8.19   100.0

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525 entries, 0 to 524
Data columns (total 6 columns):
nama_series     525 non-null object
kode_wilayah    525 non-null object
wilayah         525 non-null object
tahun           525 non-null object
nilai           361 non-null float64
target          15 non-null float64
dtypes: float64(2), object(4)
memory usage: 24.7+ KB


# *Checkpoint*
Sekarang kita sudah belajar cara:
- Memasukkan data csv ke dalam program kita untuk diolah lebih lanjut (`read_csv`)
- Merubah tipe kolom
- Mengakses kolom, baris tertentu sesuai perintah (`loc`, `iloc`)
- Mengeksplor data (`shape`, `info`, `describe`, `unique`, `nunique`)
- Memilih sebagian data (mengombinasikan `loc`, logika operator )


![alt text](https://media0.giphy.com/media/11sBLVxNs7v6WA/giphy.gif?cid=790b76115d44252537325577555e8107&rid=giphy.gif)
